### ISSUE TITLE
Question : How many nodes on AKS needed for persistent cross cluster raft network

### ISSUE THREAD

**Issue by ashutosh-ibm** [APGGroeiFabriek/PIVT#47]
I have created single node cluster on Azure AKS. With persistence enabled I am getting following error

Type     Reason            Age        From               Message
  ----     ------            ----       ----               -------
  Warning  FailedScheduling  <unknown>  default-scheduler  error while running "VolumeBinding" filter plugin for pod "hlf-orderer--pivt--orderer0-0": pod has unbound immediate PersistentVolumeClaims
  Warning  FailedScheduling  <unknown>  default-scheduler  error while running "VolumeBinding" filter plugin for pod "hlf-orderer--pivt--orderer0-0": pod has unbound immediate PersistentVolumeClaims
  Warning  FailedScheduling  <unknown>  default-scheduler  0/1 nodes are available: 1 node(s) exceed max volume count.

I think I have hit the limit imposed by Azure for this vm image . Is there a way to deploy this with single node? How do I calculate number of nodes needed in this case?

Thanks

**Comment by raftAtGit** [comment]
On Azure AKS, different VM types has different limits for the maximum number of attached disks. You can pick a larger one.

Alternatively, you can install Minikube on a single VM. Possibly it will allow unlimited number of PVCs (I'm guessing this)

You need 2 disks per peer node (peer + CouchDB) and one disk per orderer node. Make the calculation based on this.

You can indeed, omit CouchDB disks for a bare minimum. In this case, peers will re-construct CouchDB at each restart.

**Comment by ashutoshk78** [comment]
I have installed minikube on a EC2 instance. However it is not able to get external IPs for load balancers. Any idea? I have public subnet and internet gateway . Also enabled ingress on minikube. 
Thanks

**Comment by ashutoshk78** [comment]
Also on minikube persistent volume does not work.  logs ...

  Warning  FailedScheduling  58s (x3 over 2m3s)  default-scheduler  running "VolumeBinding" filter plugin for pod "hlf-peer--karga--peer0-0": pod has unbound immediate PersistentVolumeClaims

Not sure if any configuration changes needed .

**Comment by raftAtGit** [comment]
why do you need LoadBalancer? 

for `PersistentVolumeClaims`, you might need to change [storage class name](https://github.com/APGGroeiFabriek/PIVT/blob/master/fabric-kube/hlf-kube/values.yaml#L22).

what is the output of this command:
```
kubectl get storageclass
```

**Comment by ashutoshk78** [comment]
why do you need LoadBalancer?

 kubectl --namespace two get svc -l addToExternalHostAliases=true --watch 
        This command is not able to get external ip. It remains in pending state. 

 kubectl get storageclass shows 
NAME                 PROVISIONER                RECLAIMPOLICY   VOLUMEBINDINGMODE   ALLOWVOLUMEEXPANSION   AGE
standard (default)   k8s.io/minikube-hostpath   Delete          Immediate           false                  130m

**Comment by raftAtGit** [comment]
so you need to set storage class name to `standard`

himm, I dont think you can easily setup up cross-cluster sample with minikube. `LoadBalancer` type of services are glue code, they are not part of Kubernetes, they are provided by cloud providers. dont think this is possible with minikube.

maybe one option is, change the service type of external peer/orderer services to `NodePort` and manually configure some AWS load balancers to use that NodePorts. possibly will work but not trivial..

**Comment by ashutoshk78** [comment]
I was trying cross cluster raft tls  on AWS EKS .  collect_external_host_aliases script does not return IP as there is no ip field in map for hlf-peer-ingress and hlf-orderer-ingress. Even for other SVCs external ip is not there. Do I need to modify the script?

**Comment by raftAtGit** [comment]
dont think so. what is the output of this command:
```
kubectl get svc -n kube-system | grep hlf-peer-ingress
```

**Comment by ashutoshk78** [comment]
hlf-peer-ingress-nginx-ingress-controller           LoadBalancer   10.100.164.103   a47777c7a045a45f0891f336d50bf94c-1485895916.us-east-1.elb.amazonaws.com   7051:30705/TCP   43m


hlf-peer-ingress-nginx-ingress-default-backend      ClusterIP      10.100.144.22    <none>                                                                    80/TCP           43m

**Comment by ashutoshk78** [comment]
hlf-peer-ingress-nginx-ingress-controller           LoadBalancer
10.100.164.103
a47777c7a045a45f0891f336d50bf94c-1485895916.us-east-1.elb.amazonaws.com
7051:30705/TCP   43m
hlf-peer-ingress-nginx-ingress-default-backend      ClusterIP
 10.100.144.22    <none>
                 80/TCP           43m

**Comment by raftAtGit** [comment]
and what is the output of:
```
kubectl describe svc -n kube-system hlf-peer-ingress-nginx-ingress-controller
```

paste the output between 3 ` (back quote) please

**Comment by ashutoshk78** [comment]
Name:                     hlf-peer-ingress-nginx-ingress-controller


Namespace:                kube-system


Labels:                   app=nginx-ingress


                               chart=nginx-ingress-1.38.0


                               component=controller

                              heritage=Tiller

                             release=hlf-peer-ingress

Annotations:              <none>

Selector:
app.kubernetes.io/component=controller,app=nginx-ingress,release=hlf-peer-ingress



Type:                     LoadBalancer


IP:                       10.100.190.162


LoadBalancer Ingress:     xxxx-1871078149.us-east-1.elb.amazonaws.com



 Port:                          https  7051/TCP

TargetPort:                 https/TCP

NodePort:                  https  30991/TCP

Endpoints:                 <none>

Session Affinity:         None

External Traffic Policy:  Cluster

Events:
  Type    Reason                Age   From                Message


  ----    ------                ----  ----                -------


  Normal  EnsuringLoadBalancer  47s   service-controller  Ensuring load
balancer




  Normal  EnsuredLoadBalancer   44s   service-controller  Ensured load
balancer

**Comment by ashutoshk78** [comment]
''' 
Name:                     hlf-peer-ingress-nginx-ingress-controller
Namespace:                kube-system
Labels:                   app=nginx-ingress
                          chart=nginx-ingress-1.38.0
                          component=controller
                          heritage=Tiller
                          release=hlf-peer-ingress
Annotations:              <none>
selector:                 app.kubernetes.io/component=controller,app=nginx-ingress,release=hlf-peer-ingress
Type:                     LoadBalancer
IP:                       10.100.190.162
LoadBalancer Ingress:     xxxxx.us-east-1.elb.amazonaws.com
Port:                     https  7051/TCP
TargetPort:               https/TCP
NodePort:                 https  30991/TCP
Endpoints:                <none>
Session Affinity:         None
External Traffic Policy:  Cluster
Events:
  Type    Reason                Age   From                Message
  ----    ------                ----  ----                -------
  Normal  EnsuringLoadBalancer  47s   service-controller  Ensuring load balancer
  Normal  EnsuredLoadBalancer   44s   service-controller  Ensured load balancer  '''

**Comment by raftAtGit** [comment]
well, looks like `LoadBalancer Ingress` field does not reflect the `IP` but `FQDN` of the Ingress load balancer. 

in this case, you cannot use `collect_external_host_aliases.sh` script as it is.

maybe, you can modify the script, such that, after getting FQDN, you can use AWS CLI to get the IP of the load balancer.

**Comment by ashutoshk78** [comment]
Ya . I thought of that but I think IP is not fixed for load
balancers  against FQDN. There are  ways to use static IP for LB but that
may not be optimal solution.
Let me know your thoughts.

Thanks
Ashutosh

**Comment by raftAtGit** [comment]
how dynamic are LB IP’s? do they change on their own?

I mean AWS EC2 IP’s are also dynamic if you dont use an elastic IP, but they do not change on their own if you dont stop/start the VM.

so, if LB IP’s remain stable during the life time of LB, I would go for it. o/w using static IP’s make sense

**Comment by ashutoshk78** [comment]
Is there a way to use DNS name itself instead of IP?

**Comment by raftAtGit** [comment]
no, not for the way we use it, I mean for the ‘hostAliases’ spec.

but you can for sure register your domain names and use A(lias) DNS records to point to your LB FQDN.

**Comment by sillysachin** [comment]
Bare Metal Server setup of k8s will not have default load balancer . Such a baremetal k8s cluster would have to use https://metallb.universe.tf/installation/ to solve the external ip pending issue. I have been chasing this rabbit hole for last 10 days. Now facing Argo + MetalLB probable bugs.
